src/main/java/de/zalando/aruha/nakadi/service/ClosedConnectionsCrutch.java

```diff
@@ -0,0 +1,294 @@
+        this.port = port;
+        this.meterClosed = metricRegistry.meter("nakadi.close_crutch.closed");
+        this.featureToggleService = featureToggleService;
+    }
+
+    public void listenForConnectionClose(
+            final InetAddress address,
+            final int port,
+            final BooleanSupplier onCloseListener) {
+        if (!featureToggleService.isFeatureEnabled(FEATURE_NAME)) {
+            return;
+        }
+        LOG.debug("Listening for connection to close using crutch (" + address + ":" + port + ")");
+        final Date now = new Date();
+        synchronized (this) {
+            toAdd.add(new ListenerInfo(now, new ConnectionInfo(address, port), onCloseListener));
+        }
+    }
+
+    @Scheduled(fixedDelay = 1000)
+    public void refresh() throws IOException {
+        if (!featureToggleService.isFeatureEnabled(FEATURE_NAME)) {
+            return;
+        }
+        final long selector = System.currentTimeMillis() - DELAY_MILLIS;
+        synchronized (this) {
+            final Iterator<ListenerInfo> it = toAdd.iterator();
+            while (it.hasNext()) {
+                final ListenerInfo first = it.next();
+                if (first.addedAt.getTime() >= selector) {
+                    break;
+                }
+                listeners.computeIfAbsent(first.connectionInfo, info -> new ArrayList<>()).add(first.callback);
+                it.remove();
+            }
+        }
+        final Map<ConnectionInfo, ConnectionState> currentConnections = readAllConnectionStates();
+        final Set<ConnectionInfo> keys = new HashSet<>(listeners.keySet());
+        final AtomicInteger closedCount = new AtomicInteger();
+        for (final ConnectionInfo key : keys) {
+            if (CLOSED_STATES.contains(currentConnections.getOrDefault(key, ConnectionState.TCP_CLOSE))) {
+                LOG.info("Notifying about connection close via crutch: " + key);
+                listeners.remove(key).forEach(callable -> {
+                    if (callable.getAsBoolean()) {
+                        closedCount.incrementAndGet();
+                    }
+                });
+            }
+        }
+        if (closedCount.get() > 0) {
+            meterClosed.mark(closedCount.get());
+        }
+    }
+
+    private Map<ConnectionInfo, ConnectionState> readAllConnectionStates() throws IOException {
+        final Map<ConnectionInfo, ConnectionState> result = new HashMap<>();
+        for (final File f : new File[]{new File("/proc/net/tcp"), new File("/proc/net/tcp6")}) {
+            try (FileInputStream in = new FileInputStream(f)) {
+                result.putAll(getCurrentConnections(in));
+            } catch (final FileNotFoundException e) {
+                LOG.warn("Failed to find file " + f.getName() + ", skipping");
+            }
+        }
+        return result;
+    }
+
+    Map<ConnectionInfo, ConnectionState> getCurrentConnections(final InputStream in) throws IOException {
```

Comment:
It's not that I don't like the annotation itself. I don't like the idea of testing private methods making them package-level. I think that is an indication of that the design of the class could be better.
@VisibleForTesting is better than not having it and still testing a package-level method.